package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,690 @@
+                cutoff, size, nbeadspercell,
+                self.cutoff, self.size, <ns_int> (ncoords / self.size)
+            ))
+            print("NSGrid: Size={}x{}x{}={}".format(self.ncells[XX], self.ncells[YY], self.ncells[ZZ], self.size))
+
+        self.cell_offsets[XX] = 0
+        self.cell_offsets[YY] = self.ncells[XX]
+        self.cell_offsets[ZZ] = self.ncells[XX] * self.ncells[YY]
+
+        # Allocate memory
+        self.nbeads = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size)
+        if not self.nbeads:
+            raise MemoryError("Could not allocate memory from NSGrid.nbeads ({} bits requested)".format(sizeof(ns_int) * self.size))
+        self.beadids = NULL
+        self.cellids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.ncoords)
+        if not self.cellids:
+            raise MemoryError("Could not allocate memory from NSGrid.cellids ({} bits requested)".format(sizeof(ns_int) * self.ncoords))
+        self.nbeads_per_cell = 0
+
+        for i in range(self.size):
+            self.nbeads[i] = 0
+
+    def __dealloc__(self):
+        PyMem_Free(self.nbeads)
+        PyMem_Free(self.beadids)
+        PyMem_Free(self.cellids)
+
+    cdef ns_int coord2cellid(self, rvec coord) nogil:
+        return <ns_int> (coord[ZZ] / self.cellsize[ZZ]) * (self.ncells[XX] * self.ncells[YY]) +\
+               <ns_int> (coord[YY] / self.cellsize[YY]) * self.ncells[XX] + \
+               <ns_int> (coord[XX] / self.cellsize[XX])
+
+    cdef bint cellid2cellxyz(self, ns_int cellid, ivec cellxyz) nogil:
+        if cellid < 0:
+            return False
+        if cellid >= self.size:
+            return False
+
+        cellxyz[ZZ] = <ns_int> (cellid / self.cell_offsets[ZZ])
+        cellid -= cellxyz[ZZ] * self.cell_offsets[ZZ]
+
+        cellxyz[YY] = <ns_int> (cellid / self.cell_offsets[YY])
+        cellxyz[XX] = cellid - cellxyz[YY] * self.cell_offsets[YY]
+
+        return True
+
+    cdef fill_grid(self, real[:, ::1] coords):
+        cdef ns_int i, cellindex = -1
+        cdef ns_int ncoords = coords.shape[0]
+        cdef ns_int[:] beadcounts = np.empty(self.size, dtype=np.int)
+
+        with nogil:
+            # Initialize buffers
+            for i in range(self.size):
+                beadcounts[i] = 0
+
+            # First loop: find cellindex for each bead
+            for i in range(ncoords):
+                cellindex = self.coord2cellid(&coords[i, 0])
+
+                self.nbeads[cellindex] += 1
+                self.cellids[i] = cellindex
+
+                if self.nbeads[cellindex] > self.nbeads_per_cell:
+                    self.nbeads_per_cell = self.nbeads[cellindex]
+
+        # Allocate memory
+        self.beadids = <ns_int *> PyMem_Malloc(sizeof(ns_int) * self.size * self.nbeads_per_cell) #np.empty((self.size, nbeads_max), dtype=np.int)
+        if not self.beadids:

```

Comment:
the name nbeads_per_cell is little confusing, its the maximum number of beads in any cell right?